File: code/sat_analysis/diff_trace.py
#!/usr/bin/env python3
"""diff_trace.py --- Tempest v3 OR-差分传播追踪

对每个单比特输入差分位置，追踪其通过 Phase B + andmix4×4
的传播路径，统计活跃 AND 门数量。

AND 输出模型: 保守 OR（任一输入有差分 → 输出可能有差分）
这忽略了 XOR 取消，给出的是上界（最差情况扩散）。

运行: python diff_trace.py
"""

W = 64


def popcount(x):
    return x.bit_count()


def rotl(x, r):
    return ((x << r) | (x >> (W - r))) & ((1 << W) - 1)


def phase_b_or(diff):
    """Phase B: XOR-ROT 扩散 (OR 模型)"""
    u, v, w, z = diff
    u_out = 0
    v_out = 0
    w_out = 0
    z_out = 0
    for i in range(W):
        s1 = (u >> i) & 1
        s2 = (v >> ((i - 7) % W)) & 1
        s3 = (w >> ((i - 17) % W)) & 1
        if s1 or s2 or s3:
            u_out |= (1 << i)

        s1 = (v >> i) & 1
        s2 = (w >> ((i - 11) % W)) & 1
        s3 = (z >> ((i - 23) % W)) & 1
        if s1 or s2 or s3:
            v_out |= (1 << i)

        s1 = (w >> i) & 1
        s2 = (z >> ((i - 13) % W)) & 1
        s3 = (u >> ((i - 31) % W)) & 1
        if s1 or s2 or s3:
            w_out |= (1 << i)

        s1 = (z >> i) & 1
        s2 = (u >> ((i - 17) % W)) & 1
        s3 = (v >> ((i - 7) % W)) & 1
        if s1 or s2 or s3:
            z_out |= (1 << i)

    return (u_out, v_out, w_out, z_out)


def andmix_stage_or(t, r1, r2):
    """AND-mix 单级 (OR 传播)"""
    rot1 = rotl(t, r1)
    rot2 = rotl(t, r2)

    # 活跃 AND: rot1[i] == 1 AND rot2[i] == 1
    active = rot1 & rot2
    count = popcount(active)

    # AND 输出差分 (OR 模型)
    and_out = rot1 | rot2

    # t_new = t XOR and_out
    t_new = t ^ and_out

    return t_new, count


def andmix4_or(t):
    """andmix4 完整 4 级"""
    ROTS = [(31, 53), (17, 43), (7, 23), (5, 19)]
    total = 0
    for s, (r1, r2) in enumerate(ROTS):
        t, c = andmix_stage_or(t, r1, r2)
        total += c
    return t, total


def trace_all_1bit():
    """对所有 64 个单比特位置追踪差分传播"""
    results = []

    for bit in range(W):
        # 输入: u 的单比特差分
        in_diff = (1 << bit, 0, 0, 0)

        # Phase B
        after_b = phase_b_or(in_diff)
        hw_u = popcount(after_b[0])
        hw_v = popcount(after_b[1])
        hw_w = popcount(after_b[2])
        hw_z = popcount(after_b[3])

        # andmix4
        _, and_u = andmix4_or(after_b[0]) if after_b[0] else (0, 0)
        _, and_z = andmix4_or(after_b[3]) if after_b[3] else (0, 0)
        total_and = and_u + and_z

        results.append((bit, hw_u, hw_v, hw_w, hw_z, total_and))
        print(f"bit {bit:2d}: PhaseB-> u={hw_u:2d} v={hw_v:2d} w={hw_w:2d} "
              f"z={hw_z:2d} | and u={and_u:3d} z={and_z:2d} total={total_and:3d}")

    return results


def main():
    print(f"Tempest v3 OR-差分传播追踪 (W={W})")
    print(f"{'=' * 70}")

    results = trace_all_1bit()

    totals = [r[5] for r in results]
    min_and = min(totals)
    max_and = max(totals)
    avg_and = sum(totals) / len(totals)

    min_bits = [r[0] for r in results if r[5] == min_and]
    max_bits = [r[0] for r in results if r[5] == max_and]

    print(f"\n{'=' * 70}")
    print(f"W={W} 单比特输入统计 (OR 模型)")
    print(f"{'=' * 70}")
    print(f"最少活跃 AND: {min_and} (bit {min_bits})")
    print(f"最多活跃 AND: {max_and} (bit {max_bits})")
    print(f"平均活跃 AND: {avg_and:.1f}")
    print(f"\n安全性估计 (OR 模型上界):")
    print(f"  单轮 DP ≤ 2^(-{min_and})  (最少活跃 AND)")
    print(f"  平均 DP ≈ 2^(-{avg_and:.0f})")
    print(f"  两轮 + 输出 DP ≤ 2^(-{min_and * 2 + 4})")

    # W=12 验证 (对比 MILP)
    def run_W(width):
        global W
        old = W
        W = width
        r = trace_all_1bit()
        W = old
        return min(rr[5] for rr in r)

    print(f"\n{'=' * 70}")
    print("缩减宽度对比验证 (OR 模型 vs MILP 最优)")
    min_12 = run_W(12)
    min_8 = run_W(8)
    print(f"  W=12 OR={min_12}, MILP最优=4  (OR模型高估扩散, 合理)")
    print(f"  W=8  OR={min_8},  MILP最优=8  (OR模型低估扩散! 需分析)")


if __name__ == '__main__':
    main()

File: code/sat_analysis/diff_trace.c
/* diff_trace.c --- Tempest v3 差分传播追踪器
 *
 * 对每个单比特输入差分位置，追踪其通过 Phase B + andmix4×4
 * 的传播路径，统计活跃 AND 门数量。
 *
 * AND 输出模型: 保守 OR（任一输入有差分 → 输出可能有差分）
 * 这忽略了 XOR 取消，给出的是上界（最差情况扩散）。
 *
 * 编译:
 *   gcc -O3 -o diff_trace diff_trace.c
 *   ./diff_trace
 */
#include <stdio.h>
#include <string.h>
#include <stdint.h>

#define W 64

/* ─── 差分追踪 ─── */
typedef struct {
    uint64_t bits[4];  /* u, v, w, z */
} DiffState;

static inline int get_bit(const DiffState *s, int word, int bit) {
    return (s->bits[word] >> (bit % W)) & 1;
}
static inline void set_bit(DiffState *s, int word, int bit) {
    s->bits[word] |= (1ULL << (bit % W));
}

static int popcount64(uint64_t x) {
    return __builtin_popcountll(x);
}

/* ─── Phase B: XOR-ROT 扩散 (OR 模型) ─── */
static void phase_b_or(const DiffState *in, DiffState *out) {
    memset(out, 0, sizeof(*out));
    /* u' = u0 XOR rotl(v0,7) XOR rotl(w0,17) */
    for (int i = 0; i < W; i++) {
        if (get_bit(in, 0, i) || get_bit(in, 1, (i - 7) % W) || get_bit(in, 2, (i - 17) % W))
            set_bit(out, 0, i);
    }
    /* v' = v0 XOR rotl(w0,11) XOR rotl(z0,23) */
    for (int i = 0; i < W; i++) {
        if (get_bit(in, 1, i) || get_bit(in, 2, (i - 11) % W) || get_bit(in, 3, (i - 23) % W))
            set_bit(out, 1, i);
    }
    /* w' = w0 XOR rotl(z0,13) XOR rotl(u0,31) */
    for (int i = 0; i < W; i++) {
        if (get_bit(in, 2, i) || get_bit(in, 3, (i - 13) % W) || get_bit(in, 0, (i - 31) % W))
            set_bit(out, 2, i);
    }
    /* z' = z0 XOR rotl(u0,17) XOR rotl(v0,7) */
    for (int i = 0; i < W; i++) {
        if (get_bit(in, 3, i) || get_bit(in, 0, (i - 17) % W) || get_bit(in, 1, (i - 7) % W))
            set_bit(out, 3, i);
    }
}

/* ─── AND-mix 单级 (OR 传播) ───
 *   输入: t (64-bit diff pattern)
 *   输出: after t ^= rotl(t,r1) & rotl(t,r2)
 *   返回: 该级活跃 AND 门数
 */
static int andmix_stage_or(uint64_t *t, int r1, int r2) {
    uint64_t rot1 = ((*t) << r1) | ((*t) >> (W - r1));
    uint64_t rot2 = ((*t) << r2) | ((*t) >> (W - r2));

    /* 活跃 AND 门: rot1[i] == 1 AND rot2[i] == 1 */
    uint64_t active = rot1 & rot2;
    int count = popcount64(active);

    /* AND 输出差分 (OR 模型) */
    uint64_t and_out = rot1 | rot2;

    /* t_new = t XOR and_out */
    *t ^= and_out;

    return count;
}

/* ─── andmix4 完整 4 级 ─── */
static int andmix4_or(uint64_t t) {
    int total = 0;
    int rots[4][2] = {{31,53},{17,43},{7,23},{5,19}};
    for (int s = 0; s < 4; s++) {
        int c = andmix_stage_or(&t, rots[s][0], rots[s][1]);
        total += c;
        printf("    级%d: +%d AND (hw=%d)\n", s, c, popcount64(t));
    }
    return total;
}

/* ─── 1-bit 输入差分完整追踪 ─── */
static int trace_1bit(void) {
    int best_and = 999, worst_and = 0, total_and = 0;
    int best_bit = -1, worst_bit = -1;

    for (int bit = 0; bit < W; bit++) {
        DiffState in, after_b;
        memset(&in, 0, sizeof(in));
        set_bit(&in, 0, bit);  /* u 的第 bit 位 */

        phase_b_or(&in, &after_b);

        /* andmix4 on u */
        printf("bit %2d: PhaseB -> u_hw=%d v_hw=%d w_hw=%d z_hw=%d ",
               bit, popcount64(after_b.bits[0]), popcount64(after_b.bits[1]),
               popcount64(after_b.bits[2]), popcount64(after_b.bits[3]));

        int and_u = (after_b.bits[0] != 0) ? andmix4_or(after_b.bits[0]) : 0;
        int and_z = (after_b.bits[3] != 0) ? andmix4_or(after_b.bits[3]) : 0;
        int total = and_u + and_z;
        printf("  total: %d\n", total);

        total_and += total;
        if (total < best_and) { best_and = total; best_bit = bit; }
        if (total > worst_and) { worst_and = total; worst_bit = bit; }
    }

    printf("\n===== W=64 单比特输入统计 =====\n");
    printf("最佳输入: bit %d → %d AND\n", best_bit, best_and);
    printf("最差输入: bit %d → %d AND\n", worst_bit, worst_and);
    printf("平均活跃 AND: %.1f\n", (double)total_and / W);
    return best_and;
}

int main(void) {
    printf("Tempest v3 差分传播追踪 (OR 传播模型)\n");
    printf("W = %d, 1-bit 输入差分\n\n", W);

    int min_and = trace_1bit();
    printf("\n结论: 1-bit 输入下最少活跃 AND = %d\n", min_and);
    printf("      单轮 DP 上界 (保守) ≤ 2^(-%d)\n", min_and);
    printf("      两轮 + 输出 DP ≤ 2^(-%d)\n", min_and * 2 + 4);
    return 0;
}
